File: soyeong/programmers-2/할인-행사.py
from collections import Counter

def solution(want, number, discount):
    answer = 0
    check = {}
    # 원하는 항목과 갯수를 하나로 zip
    for w, n in zip(want, number):
        check[w] = n

    for i in range(len(discount) - 9):  # 마지막 9일 동안은 10일간의 회원 기간으로 부족하므로 기준 날짜에서 제외
        c = Counter(discount[i:i + 10])  # 매일을 기준으로 원소 개수를 확인
        if c == check:  # 위에 정리한 정보와 일치하면 모두 할인 받을수 있는 일수 +1
            answer += 1

    return answer

File: sanghyun/programmers-2/할인 행사.java
import java.util.HashMap;
import java.util.Objects;

class Solution {

    public int solution(String[] want, int[] number, String[] discount) {
        int answer = 0;
        HashMap<String,Integer> wantMap = new HashMap<>();
        HashMap<String,Integer> discountMap =  new HashMap<>();

        for (int i = 0; i < want.length; i++) {
            wantMap.put(want[i],number[i]);
        }

        for (int i = 0; i < 10; i++) {
            if(discountMap.containsKey(discount[i])){
                discountMap.put(discount[i], discountMap.getOrDefault(discount[i],0)+1);
            }else{
                discountMap.put(discount[i],1);
            }
        }

        for (int i = 10; i <= discount.length; i++) {
            boolean cheak = true;
            for (String s : want) {
                if (!Objects.equals(wantMap.get(s), discountMap.get(s))) {
                    cheak = false;
                }
            }
            if(cheak){
                answer++;
            }
            if(i == discount.length){
                break;
            }
            discountMap.put(discount[i-10],discountMap.getOrDefault(discount[i-10],0)-1);
            discountMap.put(discount[i],discountMap.getOrDefault(discount[i],0)+1);
        }
        return answer;
    }
}
